crates/ruff_benchmark/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+pub mod test_case;
+
+use crate::test_case::{TestCase, TestCaseSpeed};
+use criterion::{black_box, BenchmarkId, Criterion, Throughput};
+use ruff::linter::lint_only;
+use ruff::settings::{flags, Settings};
+use std::fmt::{Debug, Display, Formatter};
+use std::str::FromStr;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub enum Tool {
+    Linter,
+}
+
+impl Tool {
+    pub fn all() -> impl Iterator<Item = Tool> {
+        [Tool::Linter].into_iter()
+    }
+
+    pub fn benchmark(&self, criterion: &mut Criterion, cases: &[TestCase]) {
+        match self {
+            Tool::Linter => benchmark_linter(criterion, cases),
+        }
+    }
+}
+
+fn benchmark_linter(criterion: &mut Criterion, cases: &[TestCase]) {
+    let mut group = criterion.benchmark_group("linter");
+
+    for case in cases {
+        group.throughput(Throughput::Bytes(case.code().len() as u64));
+        group.measurement_time(match case.speed() {
+            TestCaseSpeed::Fast => Duration::from_secs(5),
+            TestCaseSpeed::Normal => Duration::from_secs(10),
+            TestCaseSpeed::Slow => Duration::from_secs(20),
+        });
+        group.bench_with_input(BenchmarkId::from_parameter(case.name()), case, |b, case| {
+            b.iter(|| {
+                lint_only(
+                    case.code(),
+                    &case.path(),
+                    None,
+                    &black_box(Settings::default()),
+                    flags::Noqa::Enabled,
+                    flags::Autofix::Enabled,
+                )
```

Comment:
@charliermarsh is this the right function to benchmark for linting? Does `Settings::default` generate a reasonable default configuration for benchmarking?

Reply:
For the most part, though it enables only a small set of rules (the `E` and `F` categories). That actually might be a reasonable set to use for benchmarks, since it exercises both (1) AST traversal with semantic tracking and (2) physical line iteration. But, we could increase it via `Settings::for_rules`, which generates "default settings, but with a different set of rules". I'd be fine to start with this though.